### ISSUE TITLE
"SSLError: The read operation timed out" error when loading oembeded assets

### ISSUE THREAD

**Issue by ghing** [denverfoundation/storybase#946]
Throughout April 2014, I got a number of "Internal Server Message" reports that seemed to originate from an SSLError when fetching an oembeded reasource using the micawber library.

Unfortunately, the traceback doesn't provide insight into what asset URLs that were causing  the error.  I'm wondering if someone like Flickr or another service that we use oembed with had a problem with their oembed endpoints.  

On 2014-05-12, I tried requesting the paths that generated errors in the paths and didn't get any errors, so perhaps the upstream provider has addressed this issue.  I'm still logging this issue to capture the issue if it reoccurs.
### Paths that exhibited this error
- /en/explore/
- /es/explore/
- /en/stories/the-real-cost-of-soccer/viewer/
- /en/projects/redeveloping-the-holly-from-gang-violence-to-hope/
- /en/stories/six-word-memoirs/viewer/
- /api/0.1/stories/6e7665d876034d29b65a10b672f2e4a4/
### Traceback

```
Traceback (most recent call last):

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/django/core/handlers/base.py", line 115, in get_response
    response = callback(request, *callback_args, **callback_kwargs)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)

  File "/home/floodlight/webapps/floodlight_production/floodlight/apps/storybase_story/views.py", line 48, in dispatch
    return super(ExploreStoriesView, self).dispatch(*args, **kwargs)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/django/views/generic/base.py", line 86, in dispatch
    return handler(request, *args, **kwargs)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/django/views/generic/base.py", line 153, in get
    context = self.get_context_data(**kwargs)

  File "/home/floodlight/webapps/floodlight_production/floodlight/apps/storybase_story/views.py", line 78, in get_context_data
    to_be_serialized = resource.explore_get_data_to_be_serialized(self.request)

  File "/home/floodlight/webapps/floodlight_production/floodlight/apps/storybase_story/api.py", line 345, in explore_get_data_to_be_serialized
    bundle = self.full_dehydrate(bundle)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/tastypie/resources.py", line 838, in full_dehydrate
    bundle.data[field_name] = method(bundle)

  File "/home/floodlight/webapps/floodlight_production/floodlight/apps/storybase_story/api.py", line 171, in dehydrate_featured_asset_url
    return bundle.obj.featured_asset_thumbnail_url(include_host=False)

  File "/home/floodlight/webapps/floodlight_production/floodlight/apps/storybase_asset/models.py", line 745, in featured_asset_thumbnail_url
    url = featured_asset.get_thumbnail_url(**thumbnail_options)

  File "/home/floodlight/webapps/floodlight_production/floodlight/apps/storybase_asset/models.py", line 426, in get_thumbnail_url
    resource_data = self.get_oembed_response(self.url, **params)

  File "/home/floodlight/webapps/floodlight_production/floodlight/apps/storybase_asset/models.py", line 329, in get_oembed_response
    return cls.oembed_providers.request(url, **extra_params)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/micawber/providers.py", line 86, in inner
    return fn(self, url, **params)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/micawber/providers.py", line 113, in request
    return provider.request(url, **params)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/micawber/providers.py", line 60, in request
    response = self.fetch(endpoint_url)

  File "/home/floodlight/webapps/floodlight_production/venv/lib/python2.7/site-packages/micawber/providers.py", line 27, in fetch
    resp = urlopen(req)

  File "/usr/local/lib/python2.7/urllib2.py", line 127, in urlopen
    return _opener.open(url, data, timeout)

  File "/usr/local/lib/python2.7/urllib2.py", line 410, in open
    response = meth(req, response)

  File "/usr/local/lib/python2.7/urllib2.py", line 523, in http_response
    'http', request, response, code, msg, hdrs)

  File "/usr/local/lib/python2.7/urllib2.py", line 442, in error
    result = self._call_chain(*args)

  File "/usr/local/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)

  File "/usr/local/lib/python2.7/urllib2.py", line 629, in http_error_302
    return self.parent.open(new, timeout=req.timeout)

  File "/usr/local/lib/python2.7/urllib2.py", line 404, in open
    response = self._open(req, data)

  File "/usr/local/lib/python2.7/urllib2.py", line 422, in _open
    '_open', req)

  File "/usr/local/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)

  File "/usr/local/lib/python2.7/urllib2.py", line 1222, in https_open
    return self.do_open(httplib.HTTPSConnection, req)

  File "/usr/local/lib/python2.7/urllib2.py", line 1187, in do_open
    r = h.getresponse(buffering=True)

  File "/usr/local/lib/python2.7/httplib.py", line 1045, in getresponse
    response.begin()

  File "/usr/local/lib/python2.7/httplib.py", line 409, in begin
    version, status, reason = self._read_status()

  File "/usr/local/lib/python2.7/httplib.py", line 365, in _read_status
    line = self.fp.readline(_MAXLINE + 1)

  File "/usr/local/lib/python2.7/socket.py", line 476, in readline
    data = self._sock.recv(self._rbufsize)

  File "/usr/local/lib/python2.7/ssl.py", line 241, in recv
    return self.read(buflen)

  File "/usr/local/lib/python2.7/ssl.py", line 160, in read
    return self._sslobj.read(len)

SSLError: The read operation timed out
```

**Comment by ghing** [comment]
I'm still seeing this occasionally.  Reopening.

**Comment by ghing** [comment]
I looked up the assets from stories that were known to produce this error and tried to render the content using this snippet:

```
import csv
from ssl import SSLError

from storybase_asset.models import ExternalAsset
from storybase_story.models import Story

def write_assets_csv(assets=None):
    if assets is None:
        assets = ExternalAsset.objects.filter(stories__status='published')
    fields = ['asset_id', 'url', 'html', 'error']
    with open('external_assets__20140617.csv', 'w') as f:
        writer = csv.DictWriter(f, fields)
        writer.writeheader()
        for asset in assets:
            row = {
                'asset_id': asset.asset_id, 
                'url': asset.url,
            }
            try:
                row['html'] = asset.render_html()
            except SSLError:
                row['error'] = True
            writer.writerow(row)

slugs = [
    "26th-and-welton-before-and-after",
    "soccer-access-by-public-transportation-in-the-chil",
    "the-real-cost-of-soccer",
    "redeveloping-the-holly-from-gang-violence-to-hope",
    "six-word-memoirs",
    "the-next-chapter",
]

stories = Story.objects.filter(slug__in=slugs)
assets = ExternalAsset.objects.filter(stories__in=stories)
write_assets_csv(assets)
```

On one run, I did replicate the SSLError in my local development environment, but on subsequent runs, it didn't seem to cause a problem.  

I noticed that all the oEmbeded assets appear to be from Flickr, so the issue seems like Flickr's oEmbed endpoint is timing out.

It looks like the path forward is to catch this error and return a default value.

Also, we should cache the oembed responses for better performance.

**Comment by ghing** [comment]
This is fixed in 5d81afd.  I'll deploy this evening.

**Comment by ghing** [comment]
Deployed fix.  Closing.
